discovery/ecs/client/aws.go

```diff
@@ -0,0 +1,541 @@
+
+	instances := []*ec2.Instance{}
+	log.Debugf("Getting ec2 instances")
+
+	params := &ec2.DescribeInstancesInput{
+		InstanceIds: ec2IDs,
+	}
+
+	// Start describing instances
+	for {
+		resp, err := c.ec2Cli.DescribeInstances(params)
+		if err != nil {
+			return nil, err
+		}
+
+		for _, r := range resp.Reservations {
+			for _, i := range r.Instances {
+				instances = append(instances, i)
+			}
+		}
+
+		if resp.NextToken == nil || aws.StringValue(resp.NextToken) == "" {
+			break
+		}
+		params.NextToken = resp.NextToken
+	}
+
+	log.Debugf("Retrieved %d ec2 instances", len(instances))
+	return instances, nil
+}
+
+// Retrieve will get all the service instance calling multiple AWS APIs
+func (c *AWSRetriever) Retrieve() ([]*types.ServiceInstance, error) {
+
+	// First get the clusters and map them
+	clusters, err := c.getClusters()
+	if err != nil {
+		return nil, err
+	}
+
+	for _, cl := range clusters {
+		c.cache.SetCluster(cl)
+	}
+
+	var wg sync.WaitGroup
+	var globErr error
+	globErrMsg := "rrror from other goroutine, stopping AWS SD objects gathering: %s"
+
+	// For each cluster get all the required objects from AWS
+	wg.Add(len(clusters))
+	for _, cluster := range clusters {
+		go func(cluster ecs.Cluster) {
+			defer wg.Done()
+
+			// Get cluster container instances
+			cInstances, err := c.getContainerInstances(&cluster)
+			if err != nil {
+				globErr = err
+				return
+			}
+
+			ec2Ids := []*string{}
+			for _, ci := range cInstances {
+				c.cache.setContainerInstance(ci)
+				ec2Ids = append(ec2Ids, ci.Ec2InstanceId)
+			}
+
+			// Check if someone errored before continuing
+			if globErr != nil {
+				log.Debugf(globErrMsg, globErr)
+				return
+			}
+
+			// Get cluster ec2 instances
+			instances, err := c.getInstances(ec2Ids)
+			if err != nil {
+				globErr = err
+				return
+			}
+
+			for _, i := range instances {
+				c.cache.setInstance(i)
+			}
+
+			// Check if someone errored before continuing
+			if globErr != nil {
+				log.Debugf(globErrMsg, globErr)
+				return
+			}
+
+			// Get cluster tasks
+			ts, err := c.getTasks(&cluster)
+			if err != nil {
+				globErr = err
+				return
+			}
+
+			tDefIDs := []*string{}
+			for _, t := range ts {
+				c.cache.setTask(t)
+				tDefIDs = append(tDefIDs, t.TaskDefinitionArn)
+			}
+
+			// Check if someone errored before continuing
+			if globErr != nil {
+				log.Debugf(globErrMsg, globErr)
+				return
+			}
+
+			// Get cluster services
+			srvs, err := c.getServices(&cluster)
+			if err != nil {
+				globErr = err
+				return
+			}
+
+			for _, s := range srvs {
+				c.cache.setService(s)
+			}
+
+			// Check if someone errored before continuing
+			if globErr != nil {
+				log.Debugf(globErrMsg, globErr)
+				return
+			}
+
+			// Get task definitions
+			tds, err := c.getTaskDefinitions(tDefIDs, true)
+			if err != nil {
+				globErr = err
+				return
+			}
+
+			for _, t := range tds {
+				c.cache.setTaskDefinition(t)
+			}
+
+		}(*cluster)
```

Comment:
Yes, but having it as an argument to the goroutine function is already sufficient to resolve that problem – it would only happen if it was a `func()`. No need to pass it by value additionally.